File: project/drivers/driver/python/src/scratchbird/keepalive.py
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

"""
Connection Keepalive Module
Prevents connection timeouts by periodically validating idle connections
"""

import asyncio
import time
from dataclasses import dataclass
from typing import Callable, Dict, Optional, Set
from datetime import datetime, timedelta
import threading


@dataclass
class KeepaliveConfig:
    """Configuration for connection keepalive"""
    interval: float = 120.0  # How often to check idle connections (seconds)
    max_idle_before_check: float = 600.0  # Max idle time before validation (seconds)
    validation_timeout: float = 5.0  # Timeout for validation query (seconds)


class KeepaliveTracker:
    """Tracks activity time for a single connection"""
    
    def __init__(self, config: KeepaliveConfig):
        self.config = config
        self._last_activity = time.monotonic()
        self._lock = threading.Lock()
    
    def mark_active(self):
        """Mark the connection as active"""
        with self._lock:
            self._last_activity = time.monotonic()
    
    def needs_validation(self) -> bool:
        """Check if connection needs validation"""
        with self._lock:
            idle_time = time.monotonic() - self._last_activity
            return idle_time > self.config.max_idle_before_check
    
    def idle_duration(self) -> float:
        """Get idle duration in seconds"""
        with self._lock:
            return time.monotonic() - self._last_activity


class KeepaliveManager:
    """Manages keepalive for multiple connections"""
    
    def __init__(self, config: KeepaliveConfig = None):
        self.config = config or KeepaliveConfig()
        self._trackers: Dict[str, KeepaliveTracker] = {}
        self._pingers: Dict[str, Callable[[], bool]] = {}
        self._lock = threading.Lock()
        self._stop_event = threading.Event()
        self._thread: Optional[threading.Thread] = None
    
    def start(self):
        """Start the keepalive monitoring thread"""
        self._thread = threading.Thread(target=self._monitor_loop, daemon=True)
        self._thread.start()
    
    def stop(self):
        """Stop the keepalive monitoring thread"""
        self._stop_event.set()
        if self._thread and self._thread.is_alive():
            self._thread.join(timeout=5.0)
    
    def register(self, conn_id: str, pinger: Callable[[], bool]) -> KeepaliveTracker:
        """
        Register a connection for keepalive monitoring
        
        Args:
            conn_id: Unique connection identifier
            pinger: Function that returns True if connection is healthy
        
        Returns:
            KeepaliveTracker for the connection
        """
        tracker = KeepaliveTracker(self.config)
        
        with self._lock:
            self._trackers[conn_id] = tracker
            self._pingers[conn_id] = pinger
        
        return tracker
    
    def unregister(self, conn_id: str):
        """Unregister a connection from keepalive monitoring"""
        with self._lock:
            self._trackers.pop(conn_id, None)
            self._pingers.pop(conn_id, None)
    
    def _monitor_loop(self):
        """Background thread that periodically checks connections"""
        while not self._stop_event.is_set():
            self._check_connections()
            self._stop_event.wait(timeout=self.config.interval)
    
    def _check_connections(self):
        """Validate idle connections"""
        with self._lock:
            trackers = dict(self._trackers)
            pingers = dict(self._pingers)
        
        for conn_id, tracker in trackers.items():
            if tracker.needs_validation():
                pinger = pingers.get(conn_id)
                if pinger:
                    try:
                        # Run ping with timeout
                        healthy = self._ping_with_timeout(
                            pinger, 
                            self.config.validation_timeout
                        )
                        if healthy:
                            tracker.mark_active()
                        # If not healthy, connection will be replaced on next checkout
                    except Exception:
                        # Ping failed - connection likely dead
                        pass
    
    def _ping_with_timeout(self, pinger: Callable[[], bool], timeout: float) -> bool:
        """Run a ping function with timeout"""
        import concurrent.futures
        
        with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
            future = executor.submit(pinger)
            try:
                return future.result(timeout=timeout)
            except concurrent.futures.TimeoutError:
                return False


class AsyncKeepaliveManager:
    """Async version of KeepaliveManager"""
    
    def __init__(self, config: KeepaliveConfig = None):
        self.config = config or KeepaliveConfig()
        self._trackers: Dict[str, KeepaliveTracker] = {}
        self._pingers: Dict[str, Callable[[], asyncio.Future]] = {}
        self._lock = asyncio.Lock()
        self._task: Optional[asyncio.Task] = None
        self._stop_event = asyncio.Event()
    
    async def start(self):
        """Start the keepalive monitoring task"""
        self._task = asyncio.create_task(self._monitor_loop())
    
    async def stop(self):
        """Stop the keepalive monitoring task"""
        self._stop_event.set()
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
    
    async def register(
        self, 
        conn_id: str, 
        pinger: Callable[[], asyncio.Future]
    ) -> KeepaliveTracker:
        """Register a connection for keepalive monitoring"""
        tracker = KeepaliveTracker(self.config)
        
        async with self._lock:
            self._trackers[conn_id] = tracker
            self._pingers[conn_id] = pinger
        
        return tracker
    
    async def unregister(self, conn_id: str):
        """Unregister a connection"""
        async with self._lock:
            self._trackers.pop(conn_id, None)
            self._pingers.pop(conn_id, None)
    
    async def _monitor_loop(self):
        """Background task that periodically checks connections"""
        while not self._stop_event.is_set():
            try:
                await asyncio.wait_for(
                    self._stop_event.wait(), 
                    timeout=self.config.interval
                )
            except asyncio.TimeoutError:
                await self._check_connections()
    
    async def _check_connections(self):
        """Validate idle connections"""
        async with self._lock:
            trackers = dict(self._trackers)
            pingers = dict(self._pingers)
        
        for conn_id, tracker in trackers.items():
            if tracker.needs_validation():
                pinger = pingers.get(conn_id)
                if pinger:
                    try:
                        # Run ping with timeout
                        healthy = await asyncio.wait_for(
                            pinger(),
                            timeout=self.config.validation_timeout
                        )
                        if healthy:
                            tracker.mark_active()
                    except asyncio.TimeoutError:
                        # Ping timed out - connection likely dead
                        pass
                    except Exception:
                        # Ping failed
                        pass

File: project/drivers/driver/ruby/lib/scratchbird/keepalive.rb
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

# ScratchBird Ruby Driver
# Keepalive Manager - Prevents connection timeouts
# Copyright (c) 2025-2026 Dalton Calford

require 'timeout'

module Scratchbird
  # Configuration for keepalive
  class KeepaliveConfig
    attr_accessor :interval_ms, :max_idle_before_check_ms, :validation_timeout_ms
    
    DEFAULT_INTERVAL = 120_000           # 2 minutes
    DEFAULT_MAX_IDLE = 600_000           # 10 minutes
    DEFAULT_VALIDATION_TIMEOUT = 5_000   # 5 seconds
    
    def initialize(options = {})
      @interval_ms = options[:interval_ms] || DEFAULT_INTERVAL
      @max_idle_before_check_ms = options[:max_idle_before_check_ms] || DEFAULT_MAX_IDLE
      @validation_timeout_ms = options[:validation_timeout_ms] || DEFAULT_VALIDATION_TIMEOUT
    end
  end
  
  # Tracks activity for a single connection
  class KeepaliveTracker
    def initialize(config)
      @config = config
      @last_activity = Time.now
      @mutex = Mutex.new
    end
    
    def mark_active
      @mutex.synchronize { @last_activity = Time.now }
    end
    
    def needs_validation?
      idle_duration_ms > @config.max_idle_before_check_ms
    end
    
    def idle_duration_ms
      ((Time.now - @last_activity) * 1000).to_i
    end
  end
  
  # Manages keepalive for multiple connections
  class KeepaliveManager
    def initialize(config = KeepaliveConfig.new)
      @config = config
      @trackers = {}
      @connections = {}
      @mutex = Mutex.new
      @running = false
      @thread = nil
    end
    
    def start
      return if @running
      @running = true
      @thread = Thread.new { monitor_loop }
    end
    
    def stop
      return unless @running
      @running = false
      @thread&.join(5)
    end
    
    def register(connection_id, connection, &pinger)
      tracker = KeepaliveTracker.new(@config)
      @mutex.synchronize do
        @trackers[connection_id] = tracker
        @connections[connection_id] = { connection: connection, pinger: pinger }
      end
      tracker
    end
    
    def unregister(connection_id)
      @mutex.synchronize do
        @trackers.delete(connection_id)
        @connections.delete(connection_id)
      end
    end
    
    def monitored_count
      @mutex.synchronize { @trackers.size }
    end
    
    private
    
    def monitor_loop
      while @running
        sleep(@config.interval_ms / 1000.0)
        check_connections
      end
    end
    
    def check_connections
      trackers_snapshot = @mutex.synchronize { @trackers.dup }
      
      trackers_snapshot.each do |conn_id, tracker|
        if tracker.needs_validation?
          conn_info = @mutex.synchronize { @connections[conn_id] }
          next unless conn_info
          
          begin
            Timeout.timeout(@config.validation_timeout_ms / 1000.0) do
              is_healthy = conn_info[:pinger].call
              tracker.mark_active if is_healthy
            end
          rescue Timeout::Error, StandardError
            # Validation failed
          end
        end
      end
    end
  end
end
